programs/spring-util/actions/snapshot.cpp

```diff
@@ -18,29 +20,112 @@ using namespace eosio::chain;
+
+   chain_id_type chain_id = chain_id_type::empty_chain_id();
+   if(header.version <= 2) {
+      snapshot.read_section<genesis_state>([&]( auto &section ) {
+         genesis_state genesis;
+         section.read_row(genesis);
+         chain_id = genesis.compute_chain_id();
+      });
+   }
+   else if(header.version <= 4) {
+      snapshot.read_section<global_property_object>([&]( auto &section ) {
+         legacy::snapshot_global_property_object_v3 legacy_global_properties; //layout is same up to chain_id for v3 & v4
+         section.read_row(legacy_global_properties);
+         chain_id = legacy_global_properties.chain_id;
+      });
+   }
+   else {
+      snapshot.read_section<global_property_object>([&]( auto &section ) {
+         legacy::snapshot_global_property_object_v5 legacy_global_properties; //layout is same up to chain_id for v5+
+         section.read_row(legacy_global_properties);
+         chain_id = legacy_global_properties.chain_id;
+      });
+   }
+
+   block_id_type head_block;
+   block_timestamp_type head_block_time;
+   if(header.version <= 2) {
+      snapshot.read_section("eosio::chain::block_state", [&]( auto &section ) {
+         snapshot_detail::snapshot_block_header_state_legacy_v2 header_state;
+         section.read_row(header_state);
+         head_block = header_state.id;
+         head_block_time = header_state.header.timestamp;
+      });
+   }
+   else if(header.version <= 6) {
+      snapshot.read_section("eosio::chain::block_state", [&]( auto &section ) {
+         snapshot_detail::snapshot_block_header_state_legacy_v3 header_state;
+         section.read_row(header_state);
+         head_block = header_state.id;
+         head_block_time = header_state.header.timestamp;
+      });
+   }
+   else {
+      snapshot.read_section("eosio::chain::block_state", [&]( auto &section ) {
+         snapshot_detail::snapshot_block_state_data_v8 header_state;
+         section.read_row(header_state);
+         if (header_state.bs_l) {
+            head_block = header_state.bs_l->id;
+            head_block_time = header_state.bs_l->header.timestamp;
+         }
+         else if (header_state.bs) {
+            head_block = header_state.bs->block_id;
+            head_block_time = header_state.bs->header.timestamp;
+         }
+      });
+   }
+
+   std::cout << json::to_pretty_string(mutable_variant_object()("major_version", header.version)
+                                                               ("minor_version", 0)
+                                                               ("chain_id", chain_id)
+                                                               ("head_block_id", head_block)
+                                                               ("head_block_num", block_header::num_from_id(head_block))
+                                                               ("head_block_time", head_block_time))
+             << std::endl;
```

Comment:
I think that code should be in `snapshot.hpp` or `snapshot_detail.hpp` so that the snapshot format details are not spread out in too many places.

Reply:
It won't be too hard to create a `get_snapshot_info()` call in those files returning some sort of `snapshot_info` struct. Any more commonization between snapshot loading and getting snapshot info might be rather invasive.

Reply:
Or even just return a mutable_variant to keep it flexible.

Reply:
Yes, that's exactly what I was thinking of.